# Project 3 – SQL Data Analysis

### Prepared by
**[Julian Andres Belmonte Ortiz](https://www.linkedin.com/in/julianbel/)**

### Organization
DecodeLabs – Training Program 2026

---
## Project Objective

The objective of this project is to use SQL queries to analyze a sales dataset and extract meaningful business insights. Through the application of fundamental SQL operations, the project aims to demonstrate the ability to retrieve, filter, sort, group, and summarize data effectively.

Key tasks include writing SELECT statements to explore the dataset, using WHERE clauses to filter records, applying ORDER BY to organize results, and leveraging GROUP BY to identify patterns across different categories. Additionally, aggregate functions such as COUNT(), SUM(), and AVG() will be used to calculate key performance metrics and support data driven decision making.

By completing this project, valuable skills in SQL fundamentals, data querying, filtering, grouping, and analytical thinking will be developed, providing a strong foundation for more advanced data analysis and reporting techniques.


---

# Import Libraries

In [51]:
# Import required libraries

import pandas as pd
import sqlite3

### Data Loading and Exploration

**Libraries Imported**: The necessary libraries, `pandas` for data manipulation and `sqlite3` for database interaction, were successfully imported.

# Load Dataset

In [52]:
# Load Excel dataset

df = pd.read_excel("Cleaned_Dataset_Project1.xlsx")

# Display first rows

df.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice,CalculatedTotal
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04,2504.04


**Dataset Loading**: An Excel file named `Cleaned_Dataset_Project1.xlsx` was loaded into a pandas DataFrame (`df`). The `df.head()` output displayed the first few rows, showing columns like `OrderID`, `Date`, `Product`, `TotalPrice`, `PaymentMethod`, `OrderStatus`, and `ReferralSource`.

# Explore Dataset

In [53]:
# Dataset information

print("Dataset Shape:")
print(df.shape)

print("\nColumns:")
print(df.columns)

print("\nMissing Values:")
print(df.isnull().sum())

Dataset Shape:
(1200, 15)

Columns:
Index(['OrderID', 'Date', 'CustomerID', 'Product', 'Quantity', 'UnitPrice',
       'ShippingAddress', 'PaymentMethod', 'OrderStatus', 'TrackingNumber',
       'ItemsInCart', 'CouponCode', 'ReferralSource', 'TotalPrice',
       'CalculatedTotal'],
      dtype='object')

Missing Values:
OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
CalculatedTotal    0
dtype: int64


**Dataset Information**: The dataset contains `1200` rows and `15` columns. All columns appear to be complete with no missing values, which is excellent for analysis. The column names indicate a sales transaction dataset.

# Create SQL Database

In [54]:
# Create an in-memory SQLite database

conn = sqlite3.connect(':memory:')

# Load dataframe into SQL table

df.to_sql(
    'sales_data',
    conn,
    if_exists='replace',
    index=False
)

print("Table successfully created.")

Table successfully created.


### SQL Database Creation

**In-memory SQLite Database**: An in-memory SQLite database was created, and your `df` DataFrame was successfully loaded into a table named `sales_data`. This allows for SQL-based querying of your dataset.

# Basic SELECT Queries

Query 1 - View All Records



In [55]:
query = """
SELECT *
FROM sales_data
LIMIT 10;
"""

pd.read_sql(query, conn)

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice,CalculatedTotal
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04,2504.04
5,ORD200005,2023-10-23,C37249,Phone,2,245.86,934 Main St,Credit Card,Shipped,TRK72976927,4,SAVE10,Instagram,491.72,491.72
6,ORD200006,2025-06-17,C83492,Laptop,1,664.42,986 Main St,Gift Card,Returned,TRK96417362,6,SAVE10,Facebook,664.42,664.42
7,ORD200007,2023-05-12,C41460,Monitor,5,149.55,706 Main St,Cash,Shipped,TRK78809193,9,FREESHIP,Facebook,747.75,747.75
8,ORD200008,2025-04-02,C26817,Phone,2,134.28,904 Main St,Gift Card,Cancelled,TRK61042692,2,No Coupon,Email,268.56,268.56
9,ORD200009,2023-11-21,C31946,Desk,4,509.38,102 Main St,Credit Card,Shipped,TRK33478363,6,SAVE10,Google,2037.52,2037.52


### Basic SELECT Queries

**View All Records (LIMIT 10)**: The query `SELECT * FROM sales_data LIMIT 10;` successfully retrieved the first 10 rows of the `sales_data` table, providing a quick overview of the data structure and content within the SQL environment.

# Select Specific Columns

In [56]:
query = """
SELECT
    OrderID,
    Product,
    Quantity,
    TotalPrice
FROM sales_data
LIMIT 10;
"""

pd.read_sql(query, conn)

,OrderID,Product,Quantity,TotalPrice
0,ORD200000,Monitor,5,2853.10
1,ORD200001,Phone,2,302.70
2,ORD200002,Tablet,5,2753.40
3,ORD200003,Chair,1,273.19
4,ORD200004,Printer,4,2504.04
5,ORD200005,Phone,2,491.72
6,ORD200006,Laptop,1,664.42
7,ORD200007,Monitor,5,747.75
8,ORD200008,Phone,2,268.56
9,ORD200009,Desk,4,2037.52


**Select Specific Columns**: The query `SELECT OrderID, Product, Quantity, TotalPrice FROM sales_data LIMIT 10;` demonstrated how to select a subset of columns, showing order details such as product and total price for the first 10 records.

# WHERE Clause

Delivered Orders

In [57]:
query = """
SELECT *
FROM sales_data
WHERE OrderStatus = 'Delivered';
"""

pd.read_sql(query, conn)

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice,CalculatedTotal
0,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04,2504.04
1,ORD200015,2023-07-17,C39416,Printer,1,473.96,942 Main St,Cash,Delivered,TRK54930938,3,No Coupon,Google,473.96,473.96
2,ORD200021,2024-11-17,C99023,Monitor,4,342.95,569 Main St,Cash,Delivered,TRK52646983,6,No Coupon,Google,1371.80,1371.80
3,ORD200024,2024-10-30,C17470,Laptop,2,230.95,956 Main St,Online,Delivered,TRK28775057,3,WINTER15,Referral,461.90,461.90
4,ORD200025,2024-07-24,C42577,Monitor,1,399.98,709 Main St,Debit Card,Delivered,TRK69423327,3,WINTER15,Email,399.98,399.98
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
226,ORD201166,2023-07-31,C95568,Chair,3,324.47,971 Main St,Cash,Delivered,TRK37075292,4,WINTER15,Email,973.41,973.41
227,ORD201168,2023-04-08,C89559,Laptop,2,331.42,297 Main St,Credit Card,Delivered,TRK82838108,6,WINTER15,Google,662.84,662.84
228,ORD201171,2023-08-28,C39911,Phone,4,518.40,699 Main St,Credit Card,Delivered,TRK80431998,7,SAVE10,Facebook,2073.60,2073.60
229,ORD201187,2024-03-13,C26340,Desk,2,184.60,562 Main St,Gift Card,Delivered,TRK59962703,7,WINTER15,Email,369.20,369.20


### WHERE Clause Analysis

**Delivered Orders**: A query filtered the data for orders with `OrderStatus = 'Delivered'`. The output shows there are `231` delivered orders, providing insight into the proportion of successful deliveries.

High Value Orders

In [58]:
query = """
SELECT *
FROM sales_data
WHERE TotalPrice > 2000;
"""

pd.read_sql(query, conn)

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice,CalculatedTotal
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10,2853.10
1,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40,2753.40
2,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04,2504.04
3,ORD200009,2023-11-21,C31946,Desk,4,509.38,102 Main St,Credit Card,Shipped,TRK33478363,6,SAVE10,Google,2037.52,2037.52
4,ORD200010,2023-12-29,C43443,Tablet,5,625.97,333 Main St,Credit Card,Returned,TRK98859248,9,WINTER15,Instagram,3129.85,3129.85
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
175,ORD201155,2023-08-13,C41478,Printer,4,534.57,667 Main St,Credit Card,Delivered,TRK62068126,6,FREESHIP,Facebook,2138.28,2138.28
176,ORD201156,2023-07-19,C20512,Laptop,4,690.78,980 Main St,Credit Card,Shipped,TRK98234737,9,No Coupon,Referral,2763.12,2763.12
177,ORD201171,2023-08-28,C39911,Phone,4,518.40,699 Main St,Credit Card,Delivered,TRK80431998,7,SAVE10,Facebook,2073.60,2073.60
178,ORD201179,2023-01-08,C84630,Laptop,5,518.55,907 Main St,Credit Card,Cancelled,TRK48054997,6,WINTER15,Instagram,2592.75,2592.75


**High Value Orders**: A query filtered for orders where `TotalPrice > 2000`. The result identified `180` high-value orders, indicating a segment of customers or products contributing significantly to revenue.

# ORDER BY

Highest Sales

In [59]:
query = """
SELECT
    OrderID,
    Product,
    TotalPrice
FROM sales_data
ORDER BY TotalPrice DESC
LIMIT 10;
"""

pd.read_sql(query, conn)

,OrderID,Product,TotalPrice
0,ORD200789,Tablet,3456.40
1,ORD201122,Monitor,3390.95
2,ORD200632,Laptop,3390.80
3,ORD200469,Chair,3384.90
4,ORD200328,Tablet,3370.20
5,ORD200107,Printer,3353.75
6,ORD200326,Laptop,3352.40
7,ORD201065,Printer,3334.00
8,ORD201031,Phone,3322.55
9,ORD200463,Laptop,3313.90


### ORDER BY Analysis

**Highest Sales**: The query `ORDER BY TotalPrice DESC LIMIT 10;` successfully identified the top 10 highest sales transactions. The product `Tablet` had the single highest sale at `3456.40`, followed closely by `Monitor` and `Laptop` sales.

Lowest Sales

In [60]:
query = """
SELECT
    OrderID,
    Product,
    TotalPrice
FROM sales_data
ORDER BY TotalPrice ASC
LIMIT 10;
"""

pd.read_sql(query, conn)

,OrderID,Product,TotalPrice
0,ORD201161,Phone,11.39
1,ORD200863,Phone,14.06
2,ORD200240,Tablet,17.24
3,ORD200542,Tablet,17.98
4,ORD200336,Laptop,18.20
5,ORD200776,Tablet,21.19
6,ORD201025,Chair,23.53
7,ORD200690,Monitor,24.48
8,ORD200926,Desk,26.95
9,ORD200473,Chair,29.86


**Lowest Sales**: Conversely, the query `ORDER BY TotalPrice ASC LIMIT 10;` showed the 10 lowest sales. The lowest recorded sale was for a `Phone` at `11.39`, indicating a wide range in transaction values.

# COUNT Aggregation

Total Number of Orders

In [61]:
query = """
SELECT COUNT(*) AS TotalOrders
FROM sales_data;
"""

pd.read_sql(query, conn)

,TotalOrders
0,1200


### COUNT Aggregation

**Total Number of Orders**: The `COUNT(*)` query confirmed that there are `1200` total orders in the `sales_data` table.

Orders by Status

In [62]:
query = """
SELECT
    OrderStatus,
    COUNT(*) AS NumberOfOrders
FROM sales_data
GROUP BY OrderStatus
ORDER BY NumberOfOrders DESC;
"""

pd.read_sql(query, conn)

,OrderStatus,NumberOfOrders
0,Cancelled,250
1,Returned,247
2,Pending,237
3,Shipped,235
4,Delivered,231


**Orders by Status**: Grouping orders by `OrderStatus` revealed the distribution: `Cancelled` (250), `Returned` (247), `Pending` (237), `Shipped` (235), and `Delivered` (231). This shows that `Cancelled` orders are the most frequent, which could be an area for investigation.

# SUM Aggregation

Total Revenue

In [63]:
query = """
SELECT
    SUM(TotalPrice) AS TotalRevenue
FROM sales_data;
"""

pd.read_sql(query, conn)

,TotalRevenue
0,1264761.96


### SUM Aggregation

**Total Revenue**: The sum of `TotalPrice` across all orders revealed a total revenue of `1,264,761.96` from the sales data.

Revenue by Product

In [64]:
query = """
SELECT
    Product,
    SUM(TotalPrice) AS Revenue
FROM sales_data
GROUP BY Product
ORDER BY Revenue DESC;
"""

pd.read_sql(query, conn)

,Product,Revenue
0,Chair,195620.11
1,Printer,195612.61
2,Laptop,192126.56
3,Tablet,186568.95
4,Monitor,175651.41
5,Desk,167459.93
6,Phone,151722.39


**Revenue by Product**: This analysis, `GROUP BY Product ORDER BY Revenue DESC;`, showed that `Chair` is the top revenue-generating product with `195,620.11`, closely followed by `Printer` and `Laptop`. `Phone` generated the least revenue at `151,722.39`.

# AVG Aggregation

Average Order Value

In [65]:
query = """
SELECT
    AVG(TotalPrice) AS AverageOrderValue
FROM sales_data;
"""

pd.read_sql(query, conn)

,AverageOrderValue
0,1053.9683


### AVG Aggregation

**Average Order Value**: The overall average order value across all transactions is `1053.97`.

Average Revenue by Product

In [66]:
query = """
SELECT
    Product,
    AVG(TotalPrice) AS AverageRevenue
FROM sales_data
GROUP BY Product
ORDER BY AverageRevenue DESC;
"""

pd.read_sql(query, conn)

,Product,AverageRevenue
0,Laptop,1110.558150
1,Chair,1098.989382
2,Printer,1080.732652
3,Monitor,1077.616012
4,Tablet,1042.284637
5,Desk,985.058412
6,Phone,972.579423


**Average Revenue by Product**: When looking at average revenue per product (`GROUP BY Product ORDER BY AverageRevenue DESC;`), `Laptop` leads with an average of `1110.56`, followed by `Chair` and `Printer`. `Phone` also has the lowest average revenue per order among products.

# Advanced GROUP BY Analysis

Payment Method Performance

In [67]:
query = """
SELECT
    PaymentMethod,
    COUNT(*) AS Orders,
    SUM(TotalPrice) AS Revenue,
    AVG(TotalPrice) AS AverageOrderValue
FROM sales_data
GROUP BY PaymentMethod
ORDER BY Revenue DESC;
"""

pd.read_sql(query, conn)

,PaymentMethod,Orders,Revenue,AverageOrderValue
0,Credit Card,234,263847.63,1127.553974
1,Online,258,262442.94,1017.220698
2,Cash,246,259786.29,1056.041829
3,Gift Card,230,246323.92,1070.973565
4,Debit Card,232,232361.18,1001.556810


### Advanced GROUP BY Analysis

**Payment Method Performance**: This query grouped sales data by `PaymentMethod`. `Credit Card` is the highest revenue generator (`263,847.63`) and also has the highest average order value (`1127.55`). While `Online` payments have the most orders (`258`), their average order value is slightly lower.

Referral Source Performance

In [68]:
query = """
SELECT
    ReferralSource,
    COUNT(*) AS Orders,
    SUM(TotalPrice) AS Revenue
FROM sales_data
GROUP BY ReferralSource
ORDER BY Revenue DESC;
"""

pd.read_sql(query, conn)

,ReferralSource,Orders,Revenue
0,Instagram,259,275285.45
1,Email,250,261808.55
2,Google,241,250441.48
3,Facebook,228,250410.90
4,Referral,222,226815.58


**Referral Source Performance**: Grouping by `ReferralSource` shows that `Instagram` is the leading source in terms of revenue (`275,285.45`) and also has the most orders (`259`). `Email` and `Google` are also significant contributors.

# Key Findings

1. The dataset contains 1,200 sales transactions.

2. Revenue is concentrated in a small group of products.

3. Some payment methods generate higher average order values.

4. Delivered orders represent the largest portion of transactions.

5. Referral sources contribute differently to overall revenue.